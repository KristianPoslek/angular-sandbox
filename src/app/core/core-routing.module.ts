import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

const routes: Routes = [
    {
        path: '',
        loadChildren: '../pokemon-list/pokemon-list.module#PokemonListModule'
    },
    {
        path: 'pokemon',
        loadChildren:
            '../pokemon-details/pokemon-details.module#PokemonDetailsModule'
    }
];

@NgModule({
    imports: [RouterModule.forRoot(routes, { useHash: true })],
    exports: [RouterModule]
})
export class CoreRoutingModule {}
