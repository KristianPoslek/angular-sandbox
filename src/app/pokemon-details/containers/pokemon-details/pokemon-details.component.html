<div class="section">
    <h2 class="title is-1">Pokemon details</h2>

    <div *ngIf="pokemon">
        <div class="columns">
            <div class="column is-one-quarter">
                <section class="section">
                    <div class="card">
                        <div class="card-image">
                            <figure class="image is-4by3">
                                <img
                                    [src]="pokemon.sprites.frontDefault"
                                    [alt]="pokemon.name"
                                />
                            </figure>
                        </div>
                        <div class="card-content">
                            <div class="media">
                                <div class="media-content">
                                    <p class="title is-4">
                                        {{ pokemon.name }}
                                    </p>
                                </div>
                            </div>
                        </div>
                    </div>
                </section>
            </div>
            <div class="column">
                <section class="section">
                    <h3 class="title is-3">General info</h3>
                    <p><strong>ID:</strong> {{ pokemon.id }}</p>
                    <p><strong>Order:</strong> {{ pokemon.order }}</p>
                    <p><strong>Weight:</strong> {{ pokemon.weight }}</p>
                    <p><strong>Height:</strong> {{ pokemon.height }}</p>
                    <p>
                        <strong>Base experience:</strong>
                        {{ pokemon.baseExperience }}
                    </p>
                </section>

                <section class="section">
                    <h3 class="title is-3">Stats</h3>
                    <p *ngFor="let stat of pokemon.stats">
                        <strong>{{ stat.name }}:</strong> {{ stat.baseStat }}
                    </p>
                </section>

                <section class="section">
                    <h3 class="title is-3">Types</h3>
                    <p *ngFor="let type of pokemon.types">
                        <span class="tag is-primary">{{ type.name }} </span>
                    </p>
                </section>

                <section class="section">
                    <h3 class="title is-3">Abilities</h3>
                    <p *ngFor="let ability of pokemon.abilities">
                        <span class="tag is-primary">{{ ability.name }} </span>
                    </p>
                </section>
            </div>
        </div>
    </div>
</div>
