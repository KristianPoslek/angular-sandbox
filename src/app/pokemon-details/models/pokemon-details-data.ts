import { Pokemon } from 'src/app/shared/models/pokemon';

export interface PokemonDetailsData {
    pokemon: Pokemon;
}
