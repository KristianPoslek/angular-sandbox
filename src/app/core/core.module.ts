import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { PokemonDetailsModule } from '../pokemon-details/pokemon-details.module';
import { PokemonListModule } from '../pokemon-list/pokemon-list.module';
import { SharedModule } from '../shared/shared.module';
import { AppComponent } from './containers/app/app.component';
import { CoreRoutingModule } from './core-routing.module';
import { metaReducers } from './state/meta.reducers';

@NgModule({
    declarations: [AppComponent],
    imports: [
        BrowserModule,
        CoreRoutingModule,
        SharedModule.forRoot(),
        PokemonListModule,
        PokemonDetailsModule,
        StoreModule.forRoot({}, { metaReducers }),
        EffectsModule.forRoot([])
    ]
})
export class CoreModule {}
