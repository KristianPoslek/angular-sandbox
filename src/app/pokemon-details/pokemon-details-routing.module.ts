import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { PokemonDetailsComponent } from './containers/pokemon-details/pokemon-details.component';

export const routes: Routes = [
    {
        path: ':name',
        component: PokemonDetailsComponent
    }
];

@NgModule({
    imports: [RouterModule.forChild(routes)],
    exports: [RouterModule]
})
export class PokemonDetailsRoutingModule {}
