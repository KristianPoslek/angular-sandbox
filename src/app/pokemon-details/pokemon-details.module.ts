import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';
import { StoreModule } from '@ngrx/store';
import { PokemonDetailsComponent } from './containers/pokemon-details/pokemon-details.component';
import { PokemonDetailsRoutingModule } from './pokemon-details-routing.module';
import { DataEffects } from './state/effects/data.effects';
import { reducers } from './state/reducers';
import { PokemonDetailsStateService } from './state/services/pokemon-details-state.service';

@NgModule({
    declarations: [PokemonDetailsComponent],
    imports: [
        CommonModule,
        StoreModule.forFeature('pokemonDetails', reducers),
        EffectsModule.forFeature([DataEffects]),
        PokemonDetailsRoutingModule
    ],
    providers: [PokemonDetailsStateService]
})
export class PokemonDetailsModule {}
