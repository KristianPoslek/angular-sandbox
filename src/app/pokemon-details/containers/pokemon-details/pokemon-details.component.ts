import { Component, OnDestroy, OnInit } from '@angular/core';
import { ActivatedRoute, ParamMap } from '@angular/router';
import { Subscription } from 'rxjs';
import { Pokemon } from 'src/app/shared/models/pokemon';
import { SharedStateService } from 'src/app/shared/state/services/shared-state.service';
import { PokemonDetailsData } from '../../models/pokemon-details-data';
import { PokemonDetailsStateService } from '../../state/services/pokemon-details-state.service';

@Component({
    selector: 'pkmn-details',
    templateUrl: './pokemon-details.component.html'
})
export class PokemonDetailsComponent implements OnInit, OnDestroy {
    subscriptions: Subscription = new Subscription();
    name: string;
    pokemon: Pokemon;

    constructor(
        private route: ActivatedRoute,
        private pokemonDetailsStateService: PokemonDetailsStateService,
        private sharedStateService: SharedStateService
    ) {}

    ngOnInit() {
        this.subscriptions.add(
            this.pokemonDetailsStateService
                .getDataState()
                .subscribe((data: PokemonDetailsData) => {
                    this.pokemon = data.pokemon;
                })
        );

        this.subscriptions.add(
            this.route.paramMap.subscribe((params: ParamMap) => {
                this.name = params.get('name');

                this.fetchPokemonDetails();
                this.sharedStateService.addLoading();
            })
        );
    }

    fetchPokemonDetails() {
        this.pokemonDetailsStateService.fetchPokemonDetails(this.name);
    }

    ngOnDestroy() {
        this.subscriptions.unsubscribe();
    }
}
