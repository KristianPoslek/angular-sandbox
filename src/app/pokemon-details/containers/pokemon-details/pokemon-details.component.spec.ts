import { Injector } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ActivatedRoute } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { of, Subject } from 'rxjs';
import { SharedStateService } from 'src/app/shared/state/services/shared-state.service';
import { MockSharedStateService } from 'src/app/shared/state/services/shared-state.service.mock';
import { pokemonDetailsDataMock } from '../../models/pokemon-details-data.mock';
import { PokemonDetailsStateService } from '../../state/services/pokemon-details-state.service';
import { MockPokemonDetailsStateService } from '../../state/services/pokemon-details-state.service.mock';
import { PokemonDetailsComponent } from './pokemon-details.component';

describe('pokemonDetails.PokemonDetailsComponent', () => {
    let testee: PokemonDetailsComponent;
    let fixture: ComponentFixture<PokemonDetailsComponent>;
    let fixtureInjector: Injector;
    let pokemonDetailsStateServiceMock: PokemonDetailsStateService;
    let sharedStateServiceMock: SharedStateService;
    let routeMock: ActivatedRoute;
    const nameMock = 'pokemonFoo';

    beforeEach(async () => {
        TestBed.configureTestingModule({
            declarations: [PokemonDetailsComponent],
            providers: [
                {
                    provide: PokemonDetailsStateService,
                    useClass: MockPokemonDetailsStateService
                },
                {
                    provide: SharedStateService,
                    useClass: MockSharedStateService
                },
                {
                    provide: ActivatedRoute,
                    useValue: {
                        params: of({ id: 'foo' }),
                        paramMap: of({
                            get: jest.fn().mockReturnValue(nameMock)
                        })
                    }
                }
            ],
            imports: [RouterTestingModule]
        });

        fixture = TestBed.createComponent(PokemonDetailsComponent);
        testee = fixture.componentInstance;
        fixtureInjector = fixture.debugElement.injector;
        pokemonDetailsStateServiceMock = fixtureInjector.get<
            PokemonDetailsStateService
        >(PokemonDetailsStateService);
        sharedStateServiceMock = fixtureInjector.get<SharedStateService>(
            SharedStateService
        );
        routeMock = fixtureInjector.get<ActivatedRoute>(ActivatedRoute);
    });

    describe('on component init', () => {
        beforeEach(() => {
            pokemonDetailsStateServiceMock.getDataState = jest.fn(() => of());
        });

        it('should have a working snapshot', () => {
            pokemonDetailsStateServiceMock.getDataState = jest.fn(() =>
                of(pokemonDetailsDataMock)
            );
            testee.ngOnInit();
            fixture.detectChanges();

            expect(fixture).toMatchSnapshot();
        });

        it('should subscribe to pokemon details data changes', done => {
            const subject = new Subject<any>();
            pokemonDetailsStateServiceMock.getDataState = () => subject;
            testee.ngOnInit();

            pokemonDetailsStateServiceMock.getDataState().subscribe(() => {
                expect(testee.pokemon).toEqual(pokemonDetailsDataMock.pokemon);
                done();
            });

            subject.next(pokemonDetailsDataMock);
        });

        it('should handle route subscriptions', () => {
            testee.fetchPokemonDetails = jest.fn();
            testee.ngOnInit();

            expect(testee.name).toBe(nameMock);
            expect(testee.fetchPokemonDetails).toHaveBeenCalledTimes(1);
            expect(sharedStateServiceMock.addLoading).toHaveBeenCalledTimes(1);
        });
    });

    it('should enable pokemon details fetching', () => {
        testee.name = nameMock;
        testee.fetchPokemonDetails();

        expect(
            pokemonDetailsStateServiceMock.fetchPokemonDetails
        ).toHaveBeenCalledWith(nameMock);
        expect(
            pokemonDetailsStateServiceMock.fetchPokemonDetails
        ).toHaveBeenCalledTimes(1);
    });

    it('should perform on destroy actions', () => {
        testee.subscriptions.unsubscribe = jest.fn();
        testee.ngOnDestroy();

        expect(testee.subscriptions.unsubscribe).toHaveBeenCalled();
    });
});
