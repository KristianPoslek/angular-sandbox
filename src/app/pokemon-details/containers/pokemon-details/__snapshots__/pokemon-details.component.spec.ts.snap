// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pokemonDetails.PokemonDetailsComponent on component init should have a working snapshot 1`] = `
<pkmn-details
  name={[Function String]}
  pokemon={[Function Object]}
  pokemonDetailsStateService={[Function MockPokemonDetailsStateService]}
  route={[Function Object]}
  sharedStateService={[Function MockSharedStateService]}
  subscriptions={[Function Subscription]}
>
  <div
    class="section"
  >
    <h2
      class="title is-1"
    >
      Pokemon details
    </h2>
    
    <div>
      <div
        class="columns"
      >
        <div
          class="column is-one-quarter"
        >
          <section
            class="section"
          >
            <div
              class="card"
            >
              <div
                class="card-image"
              >
                <figure
                  class="image is-4by3"
                >
                  <img
                    alt="pokemonFoo"
                    src="frontDefaultFoo"
                  />
                </figure>
              </div>
              <div
                class="card-content"
              >
                <div
                  class="media"
                >
                  <div
                    class="media-content"
                  >
                    <p
                      class="title is-4"
                    >
                       pokemonFoo 
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </section>
        </div>
        <div
          class="column"
        >
          <section
            class="section"
          >
            <h3
              class="title is-3"
            >
              General info
            </h3>
            <p>
              <strong>
                ID:
              </strong>
               100
            </p>
            <p>
              <strong>
                Order:
              </strong>
               500
            </p>
            <p>
              <strong>
                Weight:
              </strong>
               200
            </p>
            <p>
              <strong>
                Height:
              </strong>
               300
            </p>
            <p>
              <strong>
                Base experience:
              </strong>
               400 
            </p>
          </section>
          <section
            class="section"
          >
            <h3
              class="title is-3"
            >
              Stats
            </h3>
            
            <p>
              <strong>
                pokemonStatNameFoo:
              </strong>
               100 
            </p>
          </section>
          <section
            class="section"
          >
            <h3
              class="title is-3"
            >
              Types
            </h3>
            
            <p>
              <span
                class="tag is-primary"
              >
                pokemonTypeNameFoo 
              </span>
            </p>
          </section>
          <section
            class="section"
          >
            <h3
              class="title is-3"
            >
              Abilities
            </h3>
            
            <p>
              <span
                class="tag is-primary"
              >
                abilityNameFoo 
              </span>
            </p>
          </section>
        </div>
      </div>
    </div>
  </div>
</pkmn-details>
`;
